doc/man/lfe_types.7
'\" t
.\" Automatically generated by Pandoc 3.5
.\"
.TH "lfe_types" "7" "2024" ""
.SH NAME
lfe_types \- LFE Types and Functions Specifications
.SH TYPES
This is a description of the type syntax.
.PP
.TS
tab(@);
l l.
T{
LFE type
T}@T{
Erlang type
T}
_
T{
\f[CR](any)\f[R]
T}@T{
\f[CR]any()\f[R]
T}
T{
\f[CR](none)\f[R]
T}@T{
\f[CR]none()\f[R]
T}
T{
\f[CR](atom)\f[R]
T}@T{
\f[CR]atom()\f[R]
T}
T{
\f[CR](integer)\f[R]
T}@T{
\f[CR]integer()\f[R]
T}
T{
\f[CR](range i1 i2)\f[R]
T}@T{
\f[CR]I1..I2\f[R]
T}
T{
\f[CR](float)\f[R]
T}@T{
\f[CR]float()\f[R]
T}
T{
\f[CR](bitstring m n)\f[R]
T}@T{
\f[CR]<<_:M,_:_*N>>\f[R]
T}
T{
\f[CR](binary)\f[R]
T}@T{
\f[CR]<<_:0,_:_*8>>\f[R]
T}
T{
\f[CR](bitstring)\f[R]
T}@T{
\f[CR]<<_:0,_:_*1>>\f[R]
T}
T{
\f[CR](nil)\f[R]
T}@T{
\f[CR][]                %% nil\f[R]
T}
T{
\f[CR]...\f[R]
T}@T{
\f[CR]...\f[R]
T}
T{
\f[CR](lambda any <type>)\f[R]
T}@T{
\f[CR]fun((...) \-> <type>)\f[R]
T}
T{
\f[CR](lambda () <type>)\f[R]
T}@T{
\f[CR]fun(() \-> <type>)\f[R]
T}
T{
\f[CR](lambda (<tlist>) <type>)\f[R]
T}@T{
\f[CR]fun((<tlist>) \-> <type>)\f[R]
T}
T{
\f[CR](map)\f[R]
T}@T{
\f[CR]map()\f[R]
T}
T{
\f[CR]#M()\f[R]
T}@T{
\f[CR]#{}\f[R]
T}
T{
\f[CR]#M(<key> <value> ...)\f[R]
T}@T{
\f[CR]#{<pairlist>}\f[R]
T}
T{
\f[CR](tuple)\f[R]
T}@T{
\f[CR]tuple()\f[R]
T}
T{
\f[CR]#()\f[R]
T}@T{
\f[CR]{}\f[R]
T}
T{
\f[CR]#(<tlist>)\f[R]
T}@T{
\f[CR]{<tlist>}\f[R]
T}
T{
\f[CR](UNION <tlist>)\f[R]
T}@T{
\f[CR]<type> | <type>\f[R]
T}
.TE
.PP
Apart from the predefined types in the Erlang type system we also have
the following predefined types which cannot be redefined:
\f[CR]UNION\f[R], \f[CR]call\f[R], \f[CR]lambda\f[R] and
\f[CR]range\f[R].
The usage of \f[CR]bitstring\f[R], \f[CR]tuple\f[R] and \f[CR]map\f[R]
have also been extended.
.PP
Note that the type \f[CR]#M()\f[R] is the empty map and the type
\f[CR]#()\f[R] is the empty tuple.
We can still use the older \f[CR](map <key valuelist>)\f[R] and
\f[CR](tuple <tlist>)\f[R] formats when declaring types for maps and
tuples.
.PP
The general form of bitstrings is \f[CR](bitstring m n)\f[R] which
denotes a bitstring which starts with \f[CR]m\f[R] bits and continues
with segments of \f[CR]n\f[R] bits.
\f[CR](binary)\f[R] is a short form for a sequence of bytes while
\f[CR](bitstring)\f[R] is a short form for a sequence of bits.
There is currently no short form for an empty binary,
\f[CR](bitstring 0 0)\f[R] must be used.
.SS Type Declarations of User\-Defined Types
\f[B]\f[CB](deftype (type\-name) type\-def)\f[B]\f[R]
.PP
\f[B]\f[CB](defopaque (type\-name) type\-def)\f[B]\f[R]
.PP
\f[B]\f[CB](deftype (type\-name par1 par2) type\-def)\f[B]\f[R]
.PP
\f[B]\f[CB](defopaque (type\-name par1 par2) type\-def)\f[B]\f[R]
.PP
For unparameterised types the parentheses around the type name are
optional.
An example:
.PP
\f[B]\f[CB](deftype foo (tuple \[aq]foo (integer) (list)))\f[B]\f[R]
.PP
\f[B]\f[CB](deftype bar (tuple \[aq]bar (integer) (list)))\f[B]\f[R]
.SS Type Information in Record Declarations
\f[B]\f[CB](defrecord rec (field1 default1 type1) (field2 default2) (field3))\f[B]\f[R]
.PP
Fields with type annotations \f[I]MUST\f[R] give a default value and
fields without type annotations get the default type \f[CR](any)\f[R].
.SH SPECIFICATIONS
.SS Type specifications of User\-Defined Functions
\f[B]\f[CB](defspec (func\-name arity) function\-spec ...)\f[B]\f[R]
.PP
where
.IP
.EX
function\-spec = (arg\-type\-list ret\-type)
function\-spec = (arg\-type\-list ret\-type constraint\-list)
function\-spec = #M(arg\-types arg\-type\-list ret\-type ret\-type)
function\-spec = #M(arg\-types arg\-type\-list ret\-type ret\-type
                   contraints constraint\-list)
arg\-type\-list = (arg\-type ...)
constraint\-list = (constraint ...)
constraint = (var var\-type)
.EE
.PP
For multiple types add more function specs.
The parentheses around the function name and the arity are optional.
For example from the docs:
.IP
.EX
(defspec foo ([(pos_integer)] (pos_integer)))

(defspec (foo 1)
  ([(pos_integer)] (pos_integer))
  ([(integer)] (integer)))

(defspec (remove\-if 2)
  ([(lambda ((any)) (boolean)) (list)] (list)))
.EE
.PP
Or with constraints:
.IP
.EX
(defspec id ((X) X ((X (tuple)))))

(defspec (foo 1)
  ([(tuple X (integer))] X ((X (atom))))
  ([(list Y)] Y ((Y (number)))))

(defspec (remove\-if 2)
  ([pred (list)] (list) [(pred (lambda ((any)) (boolean)))]))
.EE
.PP
Note that a constraint variable doesn\[cq]t need to start with an
upper\-case like an Erlang variable, though in some case it may be
easier to read.
.PP
Note we are using the alternate list form with \f[CR][ ]\f[R] instead of
parentheses to make it easier to see the function arguments.
.SH Types and function specifications in the module definition
Types can also be defined in the module declaration, for example:
.IP
.EX
(defmodule this\-module
  ...
  (type ((foo\-type) (tuple \[aq]foo (integer) (list)))
        ((bar\-type) (tuple \[aq]bar (integer) (list))))
  (spec ((foo 1) ([(integer)] (foo\-type)))
        ((bar 1) ([(atom)] (list))
                 ([(list)] (atom))))
  ...)
.EE
.SH NEW FORMS
This older style of \f[CR]defmodule\f[R] can now be written in a new
style with the type and spec forms at the top level in a more basic
form.
The examples shown in \f[CR]defmodule\f[R] can now be written at the
top\-level like:
.IP
.EX
(module this\-module)
\&...
(type foo\-type (tuple \[aq]foo (integer) (list)))
(type (bar\-type) (tuple \[aq]bar (integer) (list)))

(spec (foo 1) ([(integer)] (foo\-type)))
(spec (bar 1) ([(atom)] (list))
              ([(list)] (atom)))
\&...
.EE
.PP
These forms are equivalent to using \f[CR]deftype\f[R] and
\f[CR]defspec\f[R].
.PP
The current \f[CR]defmodule\f[R] with its definitions of types and
function specs and the current macros \f[CR]deftype\f[R],
\f[CR]defopaque\f[R] and \f[CR]defspec\f[R] are still allowed and work
as expected.
.SH AUTHORS
Robert Virding.


src/ChangeLog
2013-02-12  Robert Virding  <rv@renat.local>

    * <All files>: Changed Copyright notice and now use Apache License
    V2.0.

2011-08-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (pass_expand_expr): Only try to expand lists.

2011-07-09  Robert Virding  <rv@renat.local>

    * lfe_trans.erl (from_bitseg): Fix handling of binaries.

2011-07-02  Robert Virding  <rv@renat.local>

    * lfe_eval.erl (match_binary): Fix handling of binary and pattern
    vars.

    * lfe_lint.erl (pat_binary): Fix handling of binary and pattern
    vars.

2011-06-29  Robert Virding  <rv@renat.local>

    * lfe_lint.erl, lfe_codegen.erl, lfe_eval.erl, lfe_trans.erl: Now
    use lfe_bits to parse binaries.

2011-06-26  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec and qlc macros.

2011-06-15  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (macroexpand, macroexpand-1, macroexpand-all): No
    longer shell functions.

    * lfe_lib.erl (del_vbinding): New function.

2011-06-14  Robert Virding  <rv@renat.local>

    * lfe_io_pretty.erl (indent_type): Add funcall.

2011-06-08  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add match-spec macro.

2011-06-05  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Add simple include-lib macro.

2011-05-30  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_defrecord): Add ets/mnesia match pattern
    macro emp-XXX.

    * lfe_eval.erl, lfe_shell.erl, lfe_macro.erl: Restructure
    lfe_macro.erl and redefine interface.

2011-04-24  Robert Virding  <rv@renat.local>

    * lfe_codegen.erl (comp_gtest): Removed unnecessary special cases,
    optimizer fixes this for us.

2011-04-21  Robert Virding  <rv@renat.local>

    * lfe_lint.erl: Remove unused erroneous lfe_lib imports.

2011-04-19  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Arithmetic and comparison functions
    take multiple arguments.

2011-04-12  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (erl_comp_opts, do_ok_return): Handle
    warnings_as_errors ourselves in same way as vanilla compiler.

2011-04-08  Robert Virding  <rv@renat.local>

    * lfe_comp.erl (when_opt): Change argument order.

2011-03-16  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (expand_append): Expand ++ in such a way as to
    allow use in patterns.
    (expand_defmacro): Allow (defmacro name arg ...).

2010-12-31  Robert Virding  <rv@renat.local>

    * lfe_io_format.erl (control, newline): Handle ~n correctly.

    * lfe_io_pretty.erl (print1): Prettyprint one element lists.

2010-11-25  Robert Virding  <rv@renat>

    * lfe_lint.erl (pat_bitsegs): Use check_foldl instead.
    (gexpr_bitsegs): Use check_foreach instead.

2010-11-04  Robert Virding  <rv@renat.local>

    * lfe_macro.erl (exp_predef): Extend ? macro to handle optional
    timeout and default value.

2010-11-01  Robert Virding  <rv@renat>

    * lfe_scan.xrl: Check creating atoms/symbols and code cleanup.

2010-10-27  Robert Virding  <rv@renat.local>

    * lfe_shell.erl (prompt): Prompt adds node name when alive.

2010-09-06  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (def_record, def_rec_fields): Record macros follow
    new pattern code.
    (expand_defmacro): Add explicit list constructor for args.
    (c_l_tq): Add explicit cons constructor.

    * lfe_codegen.erl (comp_pat, pat_list): Add code to generate
    pattern code for explicit list constructors.

    * lfe_eval.erl (match, match_list): Add code to match explicit
    list constructors.

    * lfe_lint.erl (check_pat, pat_list, check_alias): Add code to
    test for explicit list constructors.

2010-08-25  Robert Virding  <rv@stanislaw.local>

    * lfe_macro.erl (c_tq, c_l_tq, c_b_tq): Add guards to generators
    in comprehensions.

    * lfe_lint.erl (format_error, check_gbody, check_gexpr)
    (gexpr_bitsegs): Better error checking in guards.

2010-07-24  Robert Virding  <rv@host-90-237-128-188.mobileonline.telia.com>

    * lfe_shell.erl (macroexpand, macroexpand_1): Return evaluated
    argument if no expansion.

2010-07-12  Robert Virding  <rv@host-90-233-145-108.mobileonline.telia.com>

    * lfe_scan.xrl: Have separate cases for integer, float and symbol
    atoms, catch errors better.

2010-06-07  Robert Virding  <rv@stanislaw.local>

    * lfe_lint.erl (expr_bitseg, gexpr_bitseg, pat_bitseg): Allow
    strings (integer lists).

    * lfe_codegen.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.erl (parse_bitsegs): Allow strings (integer lists).

    * lfe_eval.lfe (eval-gbinary): Evaluate binaries in guards.

    * lfe_eval.erl (eval_gbinary): Evaluate binaries in guards.

2010-06-03  Robert Virding  <rv@stanislaw.local>

    * All files: Update copyright to 2010.

2010-04-30  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (apply-lambda): New name and more logical argument
    order for eval-lambda/4.
    (apply-match-clauses): New name and more logical argument
    order for eval-match-clauses/3.

2010-02-24  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.erl (apply_lambda): New name and more logical argument
    order for eval_lambda/4.
    (apply_match_clauses): New name and more logical argument
    order for eval_match_clauses/3.

    * lfe_macro.erl (expand_clause, expand_ml_clause, exp_predef)
    (c_tq): Guards are now a sequence of tests (when t ...).

2010-02-20  Robert Virding  <rv@stanislaw.local>

    * lfe_io.erl (scan_and_parse): Use new re-entrant parser.

    * lfe_parse.erl: Write to become re-entrant using LL(1) grammar.

2010-02-16  Robert Virding  <rv@stanislaw.local>

    * lfe_eval.lfe (eval-guard, match-when, eval-let): Guards are now
    a sequence of tests (when t ...).

    * lfe_eval.erl (eval_guard, match_when, eval_let): Guards are now
    a sequence of tests (when t ...).

    * lfe_codegen.erl (comp_guard, comp_gtest, comp_clause_body)

src/lfe_types.erl
%% Copyright (c) 2016-2021 Robert Virding
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%% File    : lfe_types.erl
%% Author  : Robert Virding
%% Purpose : Lisp Flavoured Erlang type formatting.

%% Handling types in LFE including functions for converting between
%% Erlang and LFE type syntaxes.
%%
%% We can correctly do most types except for maps where we lose the
%% distinction between assoc and exact pairs.

-module(lfe_types).

-export([is_type_decl/1]).
-export([format_error/1]).

-export([from_type_def/1,from_type_defs/1,to_type_def/2,to_type_defs/2,
         check_type_def/3,check_type_defs/3]).

-export([from_func_spec_list/1,to_func_spec_list/2,
         check_func_spec_list/3]).

%% -compile(export_all).

-include("lfe.hrl").

%% format_error(Error) -> String.
%%  Do we really need this here?

format_error({bad_type_def,T}) ->
    lfe_io:format1("bad ~w type definition", [T]);
format_error({bad_type_syntax,T}) ->
    lfe_io:format1(<<"bad ~w type syntax">>, [T]);
format_error({bad_function_spec,S}) ->
    lfe_io:format1("bad function spec: ~w", [S]).

%% is_type_decl(Tag) -> boolean().
%%  Is Name a type declaration?

is_type_decl(type) -> true;
is_type_decl(opaque) -> true;
is_type_decl(_Other) -> false.

%% from_type_def(AST) -> Def.
%%  Translate an Erlang type definition to LFE. This takes the Erlang
%%  AST form of a type definition and translates to the LFE type
%%  syntax. No AST there of course.

%% Our special cases.
from_type_def({type,_L,union,Types}) ->         %Special case union
    ['UNION'|from_type_defs(Types)];
from_type_def({type,_L,tuple,any}) ->           %Special case tuple() -> (tuple)
    [tuple];
from_type_def({type,_L,tuple,Elems}) ->
    list_to_tuple(from_type_defs(Elems));
from_type_def({type,_L,binary,Bits}) when Bits =/= [] ->
    [bitstring|from_type_defs(Bits)];           %Flip binary<->bitstring here
%% from_type_def({type,_L,bitstring,[]}) -> [bitstring,[]];
from_type_def({type,_L,map,any}) ->             %Special case map() -> (map)
    [map];
from_type_def({type,_L,map,Pairs}) ->
    maps:from_list(from_map_pairs(Pairs));
from_type_def({type,_L1,record,[{atom,_L2,Name}|Fields]}) ->
    [record,Name|from_rec_fields(Fields)];
from_type_def({type,_L,'fun',[Args,Ret]}) ->
    [lambda,from_lambda_args(Args),from_type_def(Ret)];
%% The standard erlang types.
from_type_def({type,_L,Type,Args}) when is_list(Args) ->
    [Type|from_type_defs(Args)];
from_type_def({user_type,_L,Type,Args}) when is_list(Args) ->
    [Type|from_type_defs(Args)];
from_type_def({ann_type,_L,[_Var,Type]}) ->     %Annotated types lose variable
    from_type_def(Type);
from_type_def({remote_type,_L,[{atom,_,M},{atom,_,T},Args]}) ->
    Type = list_to_atom(lists:concat([M,":",T])),
    [Type|from_type_defs(Args)];
%% Literal values.
from_type_def({var,_L,Var}) -> Var;             %A type variable
from_type_def({atom,_L,Atom}) -> ?Q(Atom);      %Literal atom
from_type_def({integer,_L,Int}) -> Int;         %Literal integer
from_type_def({float,_L,Float}) -> Float.       %Literal float

from_type_defs(Ts) ->
    lists:map(fun from_type_def/1, Ts).

from_map_pairs(Pairs) ->
    %% Lose distinction between assoc and exact pairs.
    Fun = fun ({type,_L,_P,[Kt,Vt]}) ->
                  {from_type_def(Kt),from_type_def(Vt)}
          end,
    lists:map(Fun, Pairs).

from_rec_fields(Fields) ->
    Fun = fun ({type,_L,field_type,[{atom,_,Name},Type]}) ->
                  [Name,from_type_def(Type)]
          end,
    [ Fun(F) || F <- Fields ].

from_lambda_args({type,_L,any}) -> any;         %Any arity
from_lambda_args(Args) -> from_func_prod(Args).

%% to_type_def(Def, Line) -> AST.
%%  Translate a type definition from the LFE type syntax to the Erlang
%%  AST. We must explicitly handle our special cases.

%% Our special cases.
to_type_def(['UNION'|Types], Line) ->           %Union
    {type,Line,union,to_type_defs(Types, Line)};
to_type_def([range,I1,I2], Line) ->
    {type,Line,range,to_type_defs([I1,I2], Line)};
to_type_def([bitstring,I1,I2], Line) ->         %Flip binary<->bitstring here
    {type,Line,binary,to_type_defs([I1,I2], Line)};
to_type_def([tuple], Line) ->                   %Special case (tuple) -> tuple()
    {type,Line,tuple,any};
to_type_def([tuple|Args], Line) ->              %Not a user defined type
    {type,Line,tuple,to_type_defs(Args, Line)};
to_type_def([map], Line) ->                     %Special case (map) -> map()
    {type,Line,map,any};
to_type_def([map|Elems], Line) ->
    {type,Line,map,to_map_pairs(to_pair_list(Elems), Line)};
to_type_def([record,Name|Fields], Line) ->
    {type,Line,record,[to_lit(Name, Line)|to_rec_fields(Fields, Line)]};
to_type_def([lambda,Args,Ret], Line) ->
    {type,Line,'fun',[to_lambda_args(Args, Line),to_type_def(Ret, Line)]};
to_type_def(?Q(Val), Line) ->                   %Quoted atom literal
    to_lit(Val, Line);
to_type_def([call,?Q(M),?Q(T)|Args], Line) ->
    %% Special case mod:fun expands to (call 'mod 'fun)
    Dargs = to_type_defs(Args, Line),
    {remote_type,Line,[{atom,Line,M},{atom,Line,T},Dargs]};
%% The standard erlang types.
to_type_def([Type|Args], Line) ->
    Dargs = to_type_defs(Args, Line),
    case string:tokens(atom_to_list(Type), ":") of
        [M,T] ->                                %Remote type
            {remote_type,Line,
             [{atom,Line,list_to_atom(M)},{atom,Line,list_to_atom(T)},Dargs]};
        _ ->                                    %This will also catch a:b:c
            %% Get the right tag here.
            Tag = case erl_internal:is_type(Type, length(Args)) of
                      true -> type;
                      false -> user_type
                  end,
            {Tag,Line,Type,Dargs}
        end;
to_type_def(Tup, Line) when is_tuple(Tup) ->
    {type,Line,tuple,to_type_defs(tuple_to_list(Tup), Line)};
to_type_def(Map, Line) when ?IS_MAP(Map) ->
    ToPairs = to_map_pairs(maps:to_list(Map), Line),
    {type,Line,map,ToPairs};
to_type_def(Val, Line) when is_integer(Val) ->  %Literal integer value
    to_lit(Val, Line);
to_type_def(Val, Line) when is_float(Val) ->    %Literal float value
    to_lit(Val, Line);
to_type_def(Val, Line) when is_atom(Val) ->     %Variable
    {var,Line,Val}.

to_type_defs(Ds, Line) ->
    lists:map(fun (D) -> to_type_def(D, Line) end, Ds).

to_lit(Val, Line) when is_atom(Val) -> {atom,Line,Val};
to_lit(Val, Line) when is_integer(Val) -> {integer,Line,Val};
to_lit(Val, Line) when is_float(Val) -> {float,Line,Val}.

to_pair_list([K,V|Rest]) ->
    [{K,V}|to_pair_list(Rest)];
to_pair_list([]) -> [].

to_map_pairs(Pairs, Line) ->
    %% Have lost distinction between assoc and exact pairs.
    Fun = fun ({K,V}) ->
                  {type,Line,map_field_assoc,to_type_defs([K,V], Line)}
          end,
    [ Fun(P) || P <- Pairs ].

to_rec_fields(Fs, Line) ->
    Fun = fun ([F,Type]) ->
                  {type,Line,field_type,
                   [to_lit(F, Line),to_type_def(Type, Line)]}
          end,
    [ Fun(F) || F <- Fs ].

to_lambda_args(any, Line) -> {type,Line,any};
to_lambda_args(Args, Line) -> to_func_prod(Args, Line).

%% check_type_defs(Defs, KnownRecords, TypeVars) ->
%%     {ok,TypeVars} | {error,Error,TypeVars}.
%% check_type_def(Def, KnownRecords, TypeVars) ->
%%     {ok,TypeVars} | {error,Error,TypeVars}.
%%  Check a type definition. TypeVars is an orddict of variable names
%%  and usage counts. Errors returned are:
%%  {bad_type_syntax,Type}  - error in the type syntax
%%  {bad_type_def,Type}     - error in the type definition

%% Our special cases.
check_type_def(['UNION'|Types], Recs, Tvs) ->
    check_type_defs(Types, Recs, Tvs);
check_type_def([range,I1,I2], _Recs, Tvs) ->
    if is_integer(I1) and is_integer(I2) and (I1 =< I2) ->
            {ok,Tvs};
       true -> bad_type_syntax_error(range, Tvs)
    end;
check_type_def([tuple|Ts], Recs, Tvs) ->
    check_type_defs(Ts, Recs, Tvs);
check_type_def([bitstring,I1,I2], _Recs, Tvs) ->
    if is_integer(I1) and is_integer(I2) and (I1 >= 0) and (I2 >= 0) ->
            {ok,Tvs};
       true -> bad_type_syntax_error(bitstring, Tvs)
    end;
check_type_def([map|Pairs], Recs, Tvs) ->
    check_map_pairs(Pairs, Recs, Tvs);
check_type_def([record,Name|Fields], Recs, Tvs) ->
    check_record(Name, Fields, Recs, Tvs);
    %% if is_atom(Name) -> check_record_fields(Fields, Recs, Tvs);
    %%    true -> bad_type_syntax_error(record, Tvs)
    %% end;
check_type_def([lambda,Args,Ret], Recs, Tvs0) ->
    case check_lambda_args(Args, Recs, Tvs0) of
        {ok,Tvs1} -> check_type_def(Ret, Recs, Tvs1);
        Error -> Error
    end;
check_type_def(?Q(Val), _Recs, Tvs) -> check_type_lit(Val, Tvs);
check_type_def([call,?Q(M),?Q(T)|Args], Recs, Tvs) when is_atom(M), is_atom(T) ->
    check_type_defs(Args, Recs, Tvs);
%% The standard Erlang types.
check_type_def([Type|Args], Recs, Tvs0) when is_atom(Type) ->
    check_type_defs(Args, Recs, Tvs0);
%% Only literal tuples,  maps, integers and atoms (type variables) left now.
check_type_def(Tup, Recs, Tvs) when is_tuple(Tup) ->
    check_type_defs(tuple_to_list(Tup), Recs, Tvs);
check_type_def(Map, Recs, Tvs) when ?IS_MAP(Map) ->
    ToPairs = fun ({K,V}) -> [K,V] end,           %Convert to list pairs
    check_map_pairs(lists:flatmap(ToPairs, maps:to_list(Map)), Recs, Tvs);
check_type_def(Val, _Recs, Tvs) when is_integer(Val) -> {ok,Tvs};
check_type_def(Val, _Recs, Tvs) when is_atom(Val) ->
    %% It's a type variable.
    {ok,orddict:update_counter(Val, 1, Tvs)};
check_type_def(Def, _Recs, Tvs) ->
    bad_type_def_error(Def, Tvs).

check_type_defs(Defs, Recs, Tvs) ->
    check_type_list(fun check_type_def/3, Defs, Recs, Tvs).

check_type_lit(Val, Tvs) when is_integer(Val) ; is_atom(Val) -> {ok,Tvs};
check_type_lit(Val, Tvs) -> bad_type_def_error(Val, Tvs).

check_map_pairs([K,V|Pairs], Recs, Tvs0) ->
    case check_map_pair(K, V, Recs, Tvs0) of
        {ok,Tvs1} ->
            check_map_pairs(Pairs, Recs, Tvs1);
        Error -> Error
    end;
check_map_pairs([], _Recs, Tvs) -> {ok,Tvs};
check_map_pairs(_Other, _Recs, Tvs) ->
    bad_type_syntax_error(map, Tvs).

check_map_pair(K, V, Recs, Tvs0) ->
    case check_type_def(K, Recs, Tvs0) of
        {ok,Tvs1} -> check_type_def(V, Recs, Tvs1);
        Error -> Error
    end.

%% check_record(Record, Fields, KnownRecords, TypeVars) ->
%%     {ok,TypeVars} | {error,Error,TypeVars}.

check_record(Name, Fields, Recs, Tvs) ->
    case orddict:is_key(Name, Recs) of
        true ->
            check_record_fields(Fields, Recs, Tvs);
        false ->
            if is_atom(Name) ->
                    undefined_record_error(Name, Tvs);
               true -> bad_type_syntax_error(record, Tvs)
            end
    end.

check_record_fields(Fs, Recs, Tvs) ->
    check_type_list(fun check_record_field/3, Fs, Recs, Tvs).

check_record_field([F,T], Recs, Tvs) when is_atom(F) ->
    check_type_def(T, Recs, Tvs);
check_record_field(Other, _Recs, Tvs) ->
    bad_type_def_error(Other, Tvs).

check_lambda_args(any, _Recs, Tvs) -> {ok,Tvs};
check_lambda_args(Args, Recs, Tvs) ->
    check_type_defs(Args, Recs, Tvs).

check_type_list(Check, [E|Es], Recs, Tvs0) ->
    case Check(E, Recs, Tvs0) of
        {ok,Tvs1} -> check_type_list(Check, Es, Recs, Tvs1);
        Error -> Error
    end;
check_type_list(_Check, [], _Recs, Tvs) -> {ok,Tvs};
check_type_list(_Check, Other, _Recs, Tvs) ->     %Not a proper list
    bad_type_def_error(Other, Tvs).

%% from_func_spec_list([FuncType]) -> Type.

from_func_spec_list(Ss) ->
    Fun = fun ({type,_L,'fun',_}=Type) ->
                  from_func_spec(Type) ++ [[]];
              ({type,_L,bounded_fun,[Fun,Cs]}) ->
                  from_func_spec(Fun) ++ [from_func_constraints(Cs)]
          end,
    lists:map(Fun, Ss).

from_func_spec({type,_L,'fun',[Prod,Ret]}) ->
    [from_func_prod(Prod),from_type_def(Ret)].

from_func_prod({type,_L,product,Args}) when is_list(Args) ->
    from_type_defs(Args).                       %Function arguments

from_func_constraint({type,_,constraint,[{atom,_,is_subtype},St]}) ->
    from_subtype(St).

from_func_constraints(Cs) ->
    lists:map(fun from_func_constraint/1, Cs).

from_subtype([{var,_,Var},Type]) -> [Var,from_type_def(Type)].

%% to_func_spec_list(Specs, Line) -> AST.
%%  Take a list of function type specs and generate the ASTs for them.

to_func_spec_list(Specs, Line) ->
    %% io:format("ty ~p\n", [{tfs,Specs,Line}]),
    lists:map(fun (Spec) -> to_func_spec(Spec, Line) end, Specs).

to_func_spec([Prod,Ret], Line) ->
    to_func_spec(Prod, Ret, [], Line);
to_func_spec([Prod,Ret,Cs], Line) ->            %Future proof
    to_func_spec(Prod, Ret, Cs, Line);
to_func_spec(#{'arg-types' := Prod, 'ret-type' := Ret}=Spec, Line) ->
    Cs = maps:get(constraints, Spec, []),
    to_func_spec(Prod, Ret, Cs, Line).

to_func_spec(Prod, Ret, [], Line) ->
    {type,Line,'fun',[to_func_prod(Prod, Line),to_type_def(Ret, Line)]};
to_func_spec(Prod, Ret, Cs, Line) ->
    Fun = to_func_spec(Prod, Ret, [], Line),
    Constr = to_func_constraints(Cs, Line),
    {type,Line,bounded_fun,[Fun,Constr]}.

to_func_prod(Args, Line) ->
    {type,Line,product,to_type_defs(Args, Line)}.

to_func_constraints(Cs, Line) ->
    [ to_func_constraint(C, Line) || C <- Cs ].

to_func_constraint([Var,Type], Line) ->
    {type,Line,constraint,[{atom,Line,is_subtype},
                           [{var,Line,Var},to_type_def(Type, Line)]]}.

%% check_func_spec_list([FuncType], Arity, KnownRecords) ->
%%     {ok,[TypeVars]} | {error,Error,[TypeVars]}.
%% check_func_spec(FuncType, Arity, KnownRecords) ->
%%     {ok,TypeVars} | {error,Error,TypeVars}.
%%  Check a list of function specs. TypeVars is an orddict of variable
%%  names and usage counts. Errors returned are:
%%  {bad_function_spec,Spec}     - error in the type definition

check_func_spec_list(Ss, Ar, Recs) ->
    %% io:format("ty ~p\n", [{cfsl,Ss}]),
    check_spec_list(fun check_func_spec/3, Ss, Ar, Recs).

check_func_spec([Prod,Ret], Ar, Recs) ->
    check_func_spec(Prod, Ret, [], Ar, Recs);
check_func_spec([Prod,Ret,Cs], Ar, Recs) ->
    check_func_spec(Prod, Ret, Cs, Ar, Recs);
check_func_spec(#{'arg-types' := Prod, 'ret-type' := Ret}=Spec, Ar, Recs) ->
    Constr = maps:get(constraints, Spec, []),
    check_func_spec(Prod, Ret, Constr, Ar, Recs);
check_func_spec(Other, _Ar, _Recs) ->
    %% io:format("ty ~p\n", [{cfs,Other,_Ar}]),
    bad_function_spec_error(Other, []).

check_func_spec(Prod, Ret, Cs, Ar, Recs) ->
    Tvs0 = [],
    case check_func_prod(Prod, Ar, Recs, Tvs0) of
        {ok,Tvs1} ->
            case check_type_def(Ret, Recs, Tvs1) of
                {ok,Tvs2} ->
                    check_func_constraints(Cs, Recs, Tvs2);
                Error -> Error
            end;
        Error -> Error
    end.

check_func_prod(Args, Ar, Recs, Tvs0) ->
    %% This checks both the list and the types.
    case check_type_defs(Args, Recs, Tvs0) of
        {ok,Tvs1} ->
            if length(Args) =:= Ar -> {ok,Tvs1};
               true -> bad_function_spec_error(Args, Tvs1)
            end;
        Error -> Error
    end.

check_func_constraints([[Var,Type]|Cs], Recs, Tvs0) when is_atom(Var) ->
    Tvs1 = orddict:update_counter(Var, 1, Tvs0),
    case check_type_def(Type, Recs, Tvs1) of
        {ok,Tvs2} -> check_func_constraints(Cs, Recs, Tvs2);
        Error -> Error
    end;
check_func_constraints([], _Recs, Tvs) -> {ok,Tvs};
check_func_constraints(Other, _Recs, Tvs) ->
    bad_function_spec_error(Other, Tvs).

check_spec_list(Check, Es, Ar, Recs) ->
    check_spec_list(Check, Es, Ar, Recs, []).

check_spec_list(Check, [E|Es], Ar, Recs, Tvss) ->
    case Check(E, Ar, Recs) of
        {ok,Tvs} ->
            check_spec_list(Check, Es, Ar, Recs, Tvss ++ [Tvs]);
        Error -> Error
    end;
check_spec_list(_Check, [], _Ar, _Recs, Tvss) -> {ok,Tvss};
check_spec_list(_Check, Other, _Ar, _Recs, Tvss) ->
    %% Not a proper list.
    bad_function_spec_error(Other, Tvss).

%% Return errors.

bad_function_spec_error(Val, Tvs) -> {error,{bad_function_spec,Val},Tvs}.

bad_type_def_error(Type, Tvs) -> {error,{bad_type_def,Type},Tvs}.

bad_type_syntax_error(Type, Tvs) -> {error,{bad_type_syntax,Type},Tvs}.

undefined_record_error(Rec, Tvs) -> {error,{undefined_record,Rec},Tvs}.
